### PR TITLE
Add HTTP/1.1 Connection: TE support

### PR DESCRIPTION
**Pull request by vetinari**

### CODE REVIEW

File: compress_test.go [hunk 1]
```diff
@@ -63,3 +80,45 @@ func TestCompressHandlerGzipDeflate(t *testing.T) {
 		t.Fatalf("wrong content type, got %s want %s", w.HeaderMap.Get("Content-Type"), "text/plain; charset=utf-8")
 	}
 }
+
+func TestCompressHandlerGzip_http11(t *testing.T) {
```
Comment (kisielk) [inline compress_test.go]: s/_http11/HTTP11/

File: compress.go [hunk 2]
```diff
@@ -35,50 +35,73 @@ func (w *compressResponseWriter) Write(b []byte) (int, error) {
...
-				w.Header().Set("Content-Encoding", "gzip")
-				w.Header().Add("Vary", "Accept-Encoding")
-
-				gw := gzip.NewWriter(w)
-				defer gw.Close()
-
-				h, hok := w.(http.Hijacker)
-				if !hok { /* w is not Hijacker... oh well... */
-					h = nil
+		accepted_enc := ""
```
Comment (kisielk) [inline compress.go]: please call this `acceptedEnc` in accordance with Go variable naming conventions.

**Comment by kisielk** [comment]
Looks good after the naming changes are done.

**Comment by kisielk** [comment]
@elithrar can you review this before I merge it, seeing as you have some other proposed changes to the compress handler?

**Comment by elithrar** [comment]
I'm okay with the *code*: looks sound to me. Uses the TE header for properly formed HTTP/2 and HTTP/1.1 requests, and CE if HTTP/1.0.

`Transfer-Encoding` is (mostly) semantically correct—we are dynamically *transferring* as gzip—and this is why the header was added (see: [this SO answer](http://stackoverflow.com/questions/11641923/transfer-encoding-gzip-vs-content-encoding-gzip)).

However, from what I can find, this isn't the most common approach in the wild. Here's what I'm seeing as the *default* behaviours:

* nginx sets the `Content-Encoding` header
* Fastly sets `Content-Encoding` header
* Varnish's tests only test against the CE: gzip header 
* CloudFlare sets the `Content-Encoding` header
* Apache sets `Content-Encoding` header - ref: https://github.com/apache/httpd/blob/412391c3a4380bd744b1cd4013ccc16044521e7b/modules/filters/mod_deflate.c

This is (likely) because a lot of clients don't explicitly look for the TE header, and only inspect the CE header. I suspect that browser support is much better, but there's more than just browser clients out there.

As this is a potentially breaking change (in a subtle way) I'd want to hold off unless we can demonstrate otherwise.

**Comment by vetinari** [comment]
You'd better not merge this right now. I just found that the perl LWP::UserAgent fails for some reason. I works properly when the Transfer-Encoding is sent out in one header, but with this approach in two different headers (chunked, gzip) it fails.

**Comment by kisielk** [comment]
Alright. I will close it for now so it doesn't clutter the list of PRs. Feel free to submit a new one if you get things working.

**Pull request closed without merging**